dnssec-backend.go

```diff
@@ -0,0 +1,313 @@
+package rdns
+
+import (
+	"errors"
+	"log/slog"
+	"strings"
+	"time"
+
+	"github.com/miekg/dns"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	ErrResourceNotSigned    = errors.New("resource is not signed with RRSIG")
+	ErrNoResult             = errors.New("requested RR not found")
+	ErrDnskeyNotAvailable   = errors.New("DNSKEY RR does not exist")
+	ErrDsNotAvailable       = errors.New("DS RR does not exist")
+	ErrRRSigNotAvailable    = errors.New("RRSIG does not exist")
+	ErrInvalidRRsig         = errors.New("invalid RRSIG")
+	ErrForgedRRsig          = errors.New("forged RRSIG header")
+	ErrRrsigValidationError = errors.New("RR doesn't validate against RRSIG")
+	ErrRrsigValidityPeriod  = errors.New("invalid RRSIG validity period")
+	ErrUnknownDsDigestType  = errors.New("unknown DS digest type")
+	ErrDsInvalid            = errors.New("DS RR does not match DNSKEY")
+	ErrDelegationChain      = errors.New("AuthChain has no Delegations")
+)
+
+type RRSet struct {
+	RrSet []dns.RR
+	RrSig *dns.RRSIG
+}
+
+type SignedZone struct {
+	Zone         string
+	Dnskey       *RRSet
+	Ds           *RRSet
+	ParentZone   *SignedZone
+	PubKeyLookup map[uint16]*dns.DNSKEY
+}
+
+type AuthenticationChain struct {
+	DelegationChain []SignedZone
+}
+
+func (r *RRSet) isSigned() bool {
+	return r.RrSig != nil
+}
+
+func (r *RRSet) isEmpty() bool {
+	return len(r.RrSet) < 1
+}
+
+func (r *RRSet) checkHeaderIntegrity(qname string) bool {
+	return r.RrSig != nil && r.RrSig.Header().Name != qname
+}
+
+func (z *SignedZone) checkHasDnskeys() bool {
+	return len(z.Dnskey.RrSet) > 0
+}
+
+func (z SignedZone) lookupPubKey(keyTag uint16) *dns.DNSKEY {
+	return z.PubKeyLookup[keyTag]
+}
+
+func (z SignedZone) addPubKey(k *dns.DNSKEY) {
+	z.PubKeyLookup[k.KeyTag()] = k
+}
+
+func removeRRSIGs(response *dns.Msg) *dns.Msg {
+	var filteredAnswers []dns.RR
+	filteredResponse := response.Copy()
+	for _, rr := range filteredResponse.Answer {
+		if _, isRRSIG := rr.(*dns.RRSIG); !isRRSIG {
+			filteredAnswers = append(filteredAnswers, rr)
+		}
+	}
+
+	filteredResponse.Answer = filteredAnswers
+	return filteredResponse
+}
+
+func setDNSSECdo(q *dns.Msg) *dns.Msg {
+	if q.IsEdns0() == nil {
+		q.SetEdns0(1024, true)
+	}
+	q.IsEdns0().SetDo()
+	return q
+}
+
+func newQuery(qname string, qtype uint16) *dns.Msg {
+	dnsMessage := &dns.Msg{
+		MsgHdr: dns.MsgHdr{
+			RecursionDesired: true,
+		},
```

Comment:
Technically you don't need this as `SetQuestion()` below does it already (see https://github.com/miekg/dns/blob/v1.1.63/defaults.go#L35).